src/Xamarin.Android.Build.Tasks/Utilities/PreservePinvokesNativeAssemblyGenerator.cs

```diff
@@ -0,0 +1,406 @@
+	}
+
+	void AddReturnBranch (LlvmIrFunction func, LlvmIrFunctionLabelItem returnLabel)
+	{
+		var branch = new LlvmIrInstructions.Br (returnLabel);
+		func.Body.Add (branch);
+	}
+
+	LlvmIrFunctionLabelItem AddSwitchItem<T> (LlvmIrInstructions.Switch<T> sw, ulong hash, bool is64Bit, string? comment, LlvmIrFunctionLabelItem? label) where T: struct
+	{
+		if (is64Bit) {
+			return sw.Add ((T)(object)hash, dest: label, comment: comment);
+		}
+		return sw.Add ((T)(object)(uint)hash, dest: label, comment: comment);
+	}
+
+	LlvmIrFunctionAttributeSet MakeFindPinvokeAttributeSet (LlvmIrModule module)
+	{
+		var attrSet = new LlvmIrFunctionAttributeSet {
+			new MustprogressFunctionAttribute (),
+			new NofreeFunctionAttribute (),
+			new NorecurseFunctionAttribute (),
+			new NosyncFunctionAttribute (),
+			new NounwindFunctionAttribute (),
+			new WillreturnFunctionAttribute (),
+			new MemoryFunctionAttribute {
+				Default = MemoryAttributeAccessKind.Write,
+				Argmem = MemoryAttributeAccessKind.None,
+				InaccessibleMem = MemoryAttributeAccessKind.None,
+			},
+			new UwtableFunctionAttribute (),
+			new NoTrappingMathFunctionAttribute (true),
+		};
+
+		return module.AddAttributeSet (attrSet);
+	}
+
+	// Returns `true` for all p/invokes that we know are part of our set of components, otherwise returns `false`.
+	// Returning `false` merely means that the p/invoke isn't in any of BCL or our code and therefore we shouldn't
+	// care.  It doesn't mean the p/invoke will be removed in any way.
+	(bool preserve, string? componentName) MustPreserve (PinvokeScanner.PinvokeEntryInfo pinfo, ICollection<string> components)
+	{
+		if (String.Compare ("xa-internal-api", pinfo.LibraryName, StringComparison.Ordinal) == 0) {
+			return (true, null);
+		}
+
+		foreach (string component in components) {
+			// The most common pattern for the BCL - file name without extension
+			string componentName = Path.GetFileNameWithoutExtension (component);
+			if (Matches (pinfo.LibraryName, componentName)) {
+				return (true, componentName);
+			}
+
+			// If it starts with `lib`, drop the prefix
+			if (componentName.StartsWith ("lib", StringComparison.Ordinal)) {
+				if (Matches (pinfo.LibraryName, componentName.Substring (3))) {
+					return (true, componentName);
+				}
+			}
+
+			// Might require mapping of component name to a canonical one
+			if (libraryNameMap.TryGetValue (componentName, out string? mappedComponentName) && !String.IsNullOrEmpty (mappedComponentName)) {
+				if (Matches (pinfo.LibraryName, mappedComponentName)) {
+					return (true, componentName);
+				}
+			}
+
+			// Try full file name, as the last resort
+			if (Matches (pinfo.LibraryName, Path.GetFileName (component))) {
+				return (true, componentName);
+			}
+		}
+
+		return (false, null);
+
+		bool Matches (string libraryName, string componentName)
+		{
+			return String.Compare (libraryName, componentName, StringComparison.Ordinal) == 0;
+		}
```

Comment:
Do we really need the nested function, why noticing the string compare directly?

Reply:
It's the DRY pattern (one of the very few patterns I actually consciously and reflexively use :D). There's no reason to repeat that line in the 4 places where we do the same operation. It would mean less readable code and, should this code ever had a bug (I know, unlikely - but it's the pattern), more work to fix it.

Reply:
You are doing the String.Compare all over the file, if that is a concert, you can move the matches function out rather than in a nested function.

Reply:
It's a good idea for a separate PR. To make the change not just in this file, but across our code. We have plenty of places where we do similar string comparisons, it would be good to use a single method in their place.

Reply:
```suggestion
		bool Matches (string libraryName, string componentName)
		{
			return MonoAndroidHelper.StringEquals (libraryName, componentName, StringComparison.Ordinal);
		}
```